### PR TITLE
Expand module API documentation in line with recent submission to wiki.

### PR DESCRIPTION
**Pull request by rhuanjl**
Sync ChakraCore.h comments with recent submission to wiki:
https://github.com/Microsoft/ChakraCore-wiki/commit/e4aba5f23ebb2fbac99cc144aead3653d171847a

@dilijev @liminzhu

### CODE REVIEW

**Comment by boingoing** [comment]
This is pretty exhaustive, thanks a lot for your contribution here. I will look it over in detail but looks good from a glance.

File: lib/Jsrt/ChakraCore.h [hunk 1]
```diff
@@ -42,66 +48,115 @@ typedef void* JsModuleRecord;
...
+    /// <summary>
+    /// Callback for receiving notification when module is ready.
+    /// </summary>
     JsModuleHostInfo_NotifyModuleReadyCallback = 0x3,
+    /// <summary>
+    /// Callback for receiving notification to fetch a dependent module.
+    /// </summary>
     JsModuleHostInfo_FetchImportedModuleCallback = 0x4,
+    /// <summary>
+    /// Callback for receiving notification for calls to ```import()```
```
Comment (fatcerberus) [inline lib/Jsrt/ChakraCore.h]: Only for `import()` in "legacy" scripts - `import()` in module code uses the standard FIM callback.

File: lib/Jsrt/ChakraCore.h [hunk 2]
```diff
@@ -42,66 +48,115 @@ typedef void* JsModuleRecord;
...
+    /// Module source is UTF8.
+    /// </summary>
     JsParseModuleSourceFlags_DataIsUTF8 = 0x00000001
 } JsParseModuleSourceFlags;
 
+/// <summary>
+///    The types of host info that can be set on a module record with JsSetModuleHostInfo.
+/// </summary>
+/// <remarks>
+/// For more information see JsSetModuleHostInfo.
```
Comment (boingoing) [inline lib/Jsrt/ChakraCore.h]: Nitpick: Indentation here

File: lib/Jsrt/ChakraCore.h [hunk 3]
```diff
@@ -42,66 +48,115 @@ typedef void* JsModuleRecord;
...
+/// </remarks>
 typedef enum JsModuleHostInfoKind
 {
+    /// <summary>
+    /// An exception object - e.g. if the module file cannot be found.
```
Comment (boingoing) [inline lib/Jsrt/ChakraCore.h]: Nitpick: Indentation here

File: lib/Jsrt/ChakraCore.h [hunk 4]
```diff
@@ -42,66 +48,115 @@ typedef void* JsModuleRecord;
...
+    /// </summary>
     JsModuleHostInfo_Exception = 0x01,
+    /// <summary>
+    /// Host defined info.
```
Comment (boingoing) [inline lib/Jsrt/ChakraCore.h]: Nitpick: Indentation here

File: lib/Jsrt/ChakraCore.h [hunk 5]
```diff
@@ -42,66 +48,115 @@ typedef void* JsModuleRecord;
...
+    /// </summary>
     JsModuleHostInfo_HostDefined = 0x02,
+    /// <summary>
+    /// Callback for receiving notification when module is ready.
```
Comment (boingoing) [inline lib/Jsrt/ChakraCore.h]: Nitpick: Indentation here

File: lib/Jsrt/ChakraCore.h [hunk 6]
```diff
@@ -42,66 +48,115 @@ typedef void* JsModuleRecord;
...
+    /// </summary>
     JsModuleHostInfo_NotifyModuleReadyCallback = 0x3,
+    /// <summary>
+    /// Callback for receiving notification to fetch a dependent module.
```
Comment (boingoing) [inline lib/Jsrt/ChakraCore.h]: Nitpick: Indentation here

File: lib/Jsrt/ChakraCore.h [hunk 7]
```diff
@@ -42,66 +48,115 @@ typedef void* JsModuleRecord;
...
+    /// </summary>
     JsModuleHostInfo_FetchImportedModuleCallback = 0x4,
+    /// <summary>
+    /// Callback for receiving notification for calls to ```import()```
+    /// </summary>
     JsModuleHostInfo_FetchImportedModuleFromScriptCallback = 0x5,
+    /// <summary>
+    /// URL for use in error stack traces and debugging.
```
Comment (boingoing) [inline lib/Jsrt/ChakraCore.h]: Nitpick: Indentation here

File: lib/Jsrt/ChakraCore.h [hunk 8]
```diff
@@ -42,66 +48,115 @@ typedef void* JsModuleRecord;
...
 } JsModuleHostInfoKind;
 
 /// <summary>
 ///     User implemented callback to fetch additional imported modules in ES modules.
 /// </summary>
 /// <remarks>
-/// Notify the host to fetch the dependent module. This is the "import" part before HostResolveImportedModule in ES6 spec.
-/// This notifies the host that the referencing module has the specified module dependency, and the host need to retrieve the module back.
+/// The callback is invoked on the current runtime execution thread, therefore execution is blocked until 
+/// the callback completes. Notify the host to fetch the dependent module. This is the "import" part
```
Comment (boingoing) [inline lib/Jsrt/ChakraCore.h]: I know this block was not indented previously, but could you add an indent to align with the rest of the file?

File: lib/Jsrt/ChakraCore.h [hunk 9]
```diff
@@ -42,66 +48,115 @@ typedef void* JsModuleRecord;
...
 /// </returns>
 typedef JsErrorCode(CHAKRA_CALLBACK * FetchImportedModuleCallBack)(_In_ JsModuleRecord referencingModule, _In_ JsValueRef specifier, _Outptr_result_maybenull_ JsModuleRecord* dependentModuleRecord);
 
 /// <summary>
 ///     User implemented callback to fetch imported modules dynamically in scripts.
 /// </summary>
 /// <remarks>
-/// Notify the host to fetch the dependent module. This is used for the dynamic import() syntax.
-/// This notifies the host that the referencing module has the specified module dependency, and the host need to retrieve the module back.
+/// The callback is invoked on the current runtime execution thread, therefore execution is blocked untill
```
Comment (boingoing) [inline lib/Jsrt/ChakraCore.h]: Same note about indenting this block.

File: lib/Jsrt/ChakraCore.h [hunk 10]
```diff
@@ -42,66 +48,115 @@ typedef void* JsModuleRecord;
...
-/// <param name="dependentModuleRecord">The ModuleRecord of the dependent module. If the module was requested before from other source, return the
-///                           existing ModuleRecord, otherwise return a newly created ModuleRecord.</param>
+/// <param name="dwReferencingSourceContext">The referencing script context that calls import()</param>
+/// <param name="specifier">The specifier provided to the import() call.</param>
+/// <param name="dependentModuleRecord">The ModuleRecord of the dependent module. If the module was requested
+///                                     before from other source, return the existing ModuleRecord, otherwise
+///                                     return a newly created ModuleRecord.</param>
 /// <returns>
-///     true if the operation succeeded, false otherwise.
+///     Returns <c>JsNoError</c> if the operation succeeded or an error code otherwise.
```
Comment (boingoing) [inline lib/Jsrt/ChakraCore.h]: Good catch, this doc was plainly wrong before.

File: lib/Jsrt/ChakraCore.h [hunk 11]
```diff
@@ -42,66 +48,115 @@ typedef void* JsModuleRecord;
...
 /// </returns>
 typedef JsErrorCode(CHAKRA_CALLBACK * FetchImportedModuleFromScriptCallBack)(_In_ JsSourceContext dwReferencingSourceContext, _In_ JsValueRef specifier, _Outptr_result_maybenull_ JsModuleRecord* dependentModuleRecord);
 
 /// <summary>
 ///     User implemented callback to get notification when the module is ready.
 /// </summary>
 /// <remarks>
-/// Notify the host after ModuleDeclarationInstantiation step (15.2.1.1.6.4) is finished. If there was error in the process, exceptionVar
-/// holds the exception. Otherwise the referencingModule is ready and the host should schedule execution afterwards.
+/// The callback is invoked on the current runtime execution thread, therefore execution is blocked until the
```
Comment (boingoing) [inline lib/Jsrt/ChakraCore.h]: Indent this also, please.

File: lib/Jsrt/ChakraCore.h [hunk 12]
```diff
@@ -179,10 +234,10 @@ JsCreateEnhancedFunction(
...
 ///     Bootstrap the module loading process by creating a new module record.
 /// </remarks>
-/// <param name="referencingModule">The referencingModule as in HostResolveImportedModule (15.2.1.17). nullptr if this is the top level module.</param>
-/// <param name="normalizedSpecifier">The host normalized specifier. This is the key to a unique ModuleRecord.</param>
-/// <param name="moduleRecord">The new ModuleRecord created. The host should not try to call this API twice with the same normalizedSpecifier.
-///                           chakra will return an existing ModuleRecord if the specifier was passed in before.</param>
+/// <param name="referencingModule">The parent module of the new module - nullptr for a root module.</param>
+/// <param name="normalizedSpecifier">The normalized specifier for the module.</param>
+/// <param name="moduleRecord">The new module record. The host should not try to call this API twice
+///                            with the same normalizedSpecifier.</param>
```
Comment (boingoing) [inline lib/Jsrt/ChakraCore.h]: Is the behavior we're warning against undefined here or does this crash? The note before your change indicates host would just get the existing ModuleRecord back. If we really don't want host to do this, Jsrt layer could throw an error.

File: lib/Jsrt/ChakraCore.h [hunk 13]
```diff
@@ -193,16 +248,19 @@ JsInitializeModuleRecord(
     _Outptr_result_maybenull_ JsModuleRecord* moduleRecord);
 
 /// <summary>
-///     Parse the module source
+///     Parse the source for an ES module
 /// </summary>
 /// <remarks>
-/// This is basically ParseModule operation in ES6 spec. It is slightly different in that the ModuleRecord was initialized earlier, and passed in as an argument.
+/// This is basically ParseModule operation in ES6 spec. It is slightly different in that:
```
Comment (boingoing) [inline lib/Jsrt/ChakraCore.h]: Once more, please indent these.

File: lib/Jsrt/ChakraCore.h [hunk 14]
```diff
@@ -193,16 +248,19 @@ JsInitializeModuleRecord(
...
+/// b) This includes a check to see if the module being Parsed is the last module in the
+/// dependency tree. If it is it automatically triggers Module Instantiation.
 /// </remarks>
-/// <param name="requestModule">The ModuleRecord that holds the parse tree of the source code.</param>
+/// <param name="requestModule">The ModuleRecord being parsed.</param>
 /// <param name="sourceContext">A cookie identifying the script that can be used by debuggable script contexts.</param>
 /// <param name="script">The source script to be parsed, but not executed in this code.</param>
-/// <param name="scriptLength">The source length of sourceText. The input might contain embedded null.</param>
-/// <param name="sourceFlag">The type of the source code passed in. It could be UNICODE or utf8 at this time.</param>
+/// <param name="scriptLength">The length of sourceText in bytes. As the input might contain a embedded null.</param>
```
Comment (boingoing) [inline lib/Jsrt/ChakraCore.h]: Thanks for adding the "in bytes" phrase here. There is a big point of confusion between "byte count" and "character count"

File: lib/Jsrt/ChakraCore.h [hunk 15]
```diff
@@ -235,8 +295,20 @@ JsModuleEvaluation(
     _Outptr_result_maybenull_ JsValueRef* result);
 
 /// <summary>
-///     Set the host info for the specified module.
+///     Set host info for the specified module.
 /// </summary>
+/// <remarks>
+/// This is used for four things:
```
Comment (boingoing) [inline lib/Jsrt/ChakraCore.h]: Nitpick: indent

File: lib/Jsrt/ChakraCore.h (hunk 12 above)
Comment (rhuanjl) [reply lib/Jsrt/ChakraCore.h]: I'd like this to be edited to give the existing record back if you supply the same normalised specifier twice, see my suggestion #4740 

At the moment it creates a new record every time it's called - per spec one referencing module/specifier pair (i.e. one normalised specifier) should always lead to only one module record. Currently to implement this correctly the host to maintain their own map of records as this function will just make a new one whatever you supply (see the moduleRecord map in ch as an example.)

File: lib/Jsrt/ChakraCore.h (hunk 14 above)
Comment (rhuanjl) [reply lib/Jsrt/ChakraCore.h]: I can't credit for this one - I wrote this content for the wiki first before updating the header with it. The note on bytes was suggested by MSLaguana when reviewing my wiki submission.

File: lib/Jsrt/ChakraCore.h [hunk 16]
```diff
@@ -42,66 +48,115 @@ typedef void* JsModuleRecord;
 /// </remarks>
 typedef void *JsSharedArrayBufferContentHandle;
 
+/// <summary>
+///    Flags for parsing a module.
```
Comment (dilijev) [inline lib/Jsrt/ChakraCore.h]: nit: use 4 spaces to indent so that there are 8 columns before `Flags`.

File: lib/Jsrt/ChakraCore.h [hunk 17]
```diff
@@ -42,66 +48,115 @@ typedef void* JsModuleRecord;
...
+    ///     Module source is UTF8.
+    /// </summary>
     JsParseModuleSourceFlags_DataIsUTF8 = 0x00000001
 } JsParseModuleSourceFlags;
 
+/// <summary>
+///    The types of host info that can be set on a module record with JsSetModuleHostInfo.
+/// </summary>
+/// <remarks>
+///    For more information see JsSetModuleHostInfo.
```
Comment (dilijev) [inline lib/Jsrt/ChakraCore.h]: nit: 4 space indents

**Comment by dilijev** [comment]
Whatever text changes arise from this PR will need to be sync'd back to the wiki. @liminzhu do you want to work on a script to automate generating the wiki markdown from the source code?

File: lib/Jsrt/ChakraCore.h (hunk 12 above)
Comment (rhuanjl) [reply lib/Jsrt/ChakraCore.h]: To clarify the existing comment is just wrong - I'd like it to be right but that would require some work to change the API which is my suggestion I've referenced.

In the mean time editing the comment to document the actual current behaviour seemed appropriate @boingoing

Comment (boingoing) [reply lib/Jsrt/ChakraCore.h]: Yeah that's fair. Let's keep the comment as it is until we can fix #4740.

**Comment by boingoing** [comment]
I'd like to give @liminzhu a chance to take a look but baring any issues there, seems we have a consensus that this can be landed.

**Comment by liminzhu** [comment]
@rhuanjl :shipit: thanks for another awesome contribution!
@dilijev yeah that'll make a fun weekend project for me

**Comment by boingoing** [comment]
Thanks for your contribution. 👍

**Pull request merged**